File: 프로그래머스/3/150367. 표현 가능한 이진트리/표현 가능한 이진트리.cpp
#include <string>
#include <vector>

#include <iostream>
#include <stack>
#include <cmath>

using namespace std;

string tobinary(long long number) {
    stack<long long> st;
    while(number != 0) {
        st.push(number%2);
        number /= 2;
    }
    
    string bin;
    while(!st.empty()) {
        bin += to_string(st.top());
        st.pop();
    }
    return bin;
}

bool binary_search(const string& bin, int parent, int start, int end) {
    if (start > end) return true;
    
    if (start == end) {
        if (parent == 0 && bin[start] == '1') return false;
        return true;
    }
    
    int mid = (start + end) / 2;
    int cur = bin[mid] == '1' ? 1 : 0;
    
    if (parent == 0 && cur == 1) return false;
    
    bool l_sub = binary_search(bin, cur, start, mid-1);
    bool r_sub = binary_search(bin, cur, mid+1, end);
    return l_sub && r_sub;
}

// 2. 트리의 높이와 노드 개수를 구하고, 트리를 그린다.
bool calc(string bin) {
    int blength = bin.length();
    double h = floor(log(blength) / log(2.0)) + 1;
    int nodes = pow(2, h) - 1;
    
    // 완전 이진트리 만들기
    string zeros = "";
    for(int i=blength; i<nodes; i++) zeros += "0";
    
    string full_bin = zeros + bin;
    if (full_bin[full_bin.length()/2] == '0') return false;
    return binary_search(full_bin, 1, 0, nodes);
}

vector<int> solution(vector<long long> numbers) {
    
    // 1. 각 숫자를 이진수로 표현한다.
    // 2. 트리의 높이와 노드 개수를 구하고, 트리를 그린다.
    // 3. 0인 노드의 자식 노드는 1이어선 안된다.
    vector<int> answer;
    string str;
    for(long long n: numbers) {
        str = tobinary(n);
        bool result = calc(str);
        answer.push_back(result);
    }
    
    
    return answer;
}

File: 프로그래머스/3/150367. 표현 가능한 이진트리/표현 가능한 이진트리.java
import java.util.*;

class Solution {
    public int[] solution(long[] numbers) {
        
        // 1. 주어진 수를 이진수로 변환한다.
        int length = numbers.length;
        String[] bins = new String[length];
        for(int i=0; i<length; i++) {
            bins[i] = Long.toString(numbers[i], 2);
        }
        
        // 2. 포화 이진트리의 높이, 총 노드 개수를 구한다.
        // 3. 부모가 0인 노드는 반드시 0이어야 한다.
        int[] answer = new int[length];
        for(int i=0; i<length; i++) {
            answer[i] = calc(bins[i]) ? 1 : 0;
        }
        
        return answer;
    }
    
    // 2. 해당 이진수에 대해 포화 이진트리 표현 가능 여부 판단
    static boolean calc(String bin) {
        // 2-1. 트리의 높이와 노드 개수를 구한다.
        int blength = bin.length();
        int h = (int)Math.floor(Math.log(blength) / Math.log(2)) + 1;
        int nodes = (int)Math.pow(2, h) - 1;
        
        // 2-2. 빈 노드 개수만큼 앞에 0을 붙여서 포화 이진트리로 만든 다음에, 이진탐색을 돌린다.
        String zeros = "";
        for(int i=0; i<nodes-blength; i++) zeros += "0";
        
        String fullBin = zeros + bin;
        if (fullBin.charAt(fullBin.length() / 2) == '0') return false;
        return binarySearch(fullBin, 1, 0, nodes-1); // true or false
    }
    
    static boolean binarySearch(String fullBin, int parent, int start, int end) {
        if (start == end) {
            // 이제부터는 리프도 무조건 안심할 수 없음.
            // 리프가 1이면 -> 부모도 반드시 1
            // 리프가 0이면 -> 상관없음.
            if ((parent == 0) && (fullBin.charAt(start) == '1')) return false;
            else return true;
        }
        
        int mid = (start + end) / 2;
        
        // 부모 노드가 0일 경우 -> 자식 노드도 모두 0이어야 함.
        int cur = (fullBin.charAt(mid) == '1') ? 1 : 0;
        if (parent == 0 && cur == 1) return false;
        
        boolean l_sub = binarySearch(fullBin, cur, start, mid-1);
        boolean r_sub = binarySearch(fullBin, cur,  mid+1, end);
        return l_sub && r_sub;
    }
    
    
}